questionpy_server/collector/local_collector.py

```diff
@@ -0,0 +1,209 @@
+
+    @overload
+    def get(self, key: Path) -> Optional[str]:
+        """
+        Gets the hash of a package from its path.
+
+        :param key: The path of the package.
+        :return: The hash of the package.
+        """
+
+    @overload
+    def get(self, key: str) -> Optional[set[Path]]:
+        """
+        Gets the paths of a package from its hash.
+
+        :param key: The hash of the package.
+        :return: The paths of the package.
+        """
+
+    def get(self, key: Union[str, Path]) -> Union[Optional[set[Path]], Optional[str]]:
+        if isinstance(key, Path):
+            return self._map.get(key)
+
+        if isinstance(key, str):
+            return self._map_inv.get(key)
+
+        raise TypeError(f'Expected Path or str, got {type(key)}')
+
+    @overload
+    def pop(self, key: Path) -> Optional[str]:
+        """
+        Removes the package with the given path and returns its hash.
+
+        :param key: The path of the package.
+        :return: The hash of the package.
+        """
+
+    @overload
+    def pop(self, key: str) -> Optional[set[Path]]:
+        """
+        Removes all packages with the given hash and returns their paths.
+
+        :param key: The hash of the packages.
+        :return: The paths of the packages.
+        """
+
+    def pop(self, key: Union[Path, str]) -> Union[Optional[str], Optional[set[Path]]]:
+        if isinstance(key, Path):
+            # Get the hash of the package.
+            package_hash = self._map.pop(key, None)
+            if not package_hash:
+                return None
+
+            # Remove the path from the inverse map.
+            self._map_inv[package_hash].discard(key)
+
+            # If no paths are left, remove the hash from the map.
+            if not self._map_inv[package_hash]:
+                self._map_inv.pop(package_hash)
+
+            return package_hash
+
+        if isinstance(key, str):
+            paths = self._map_inv.pop(key, None)
+            if paths:
+                for path in paths:
+                    self._map.pop(path)
+            return paths
+
+        raise TypeError(f'Expected Path or str, got {type(key)}')
+
+
+class LocalCollectorEventHandler(PatternMatchingEventHandler):
+    """
+    Handles events for the local collector.
+    """
+
+    def __init__(self, local_collector: 'LocalCollector'):
+        super().__init__(patterns=['*.qpy'], ignore_directories=True, case_sensitive=True)
+        self._local_collector = local_collector
+
+    async def _push_package_async(self, path_str: str) -> None:
+        path = Path(path_str).resolve()
+
+        if path.suffix != '.qpy':
+            return
+
+        package_hash = calculate_hash(path)
+        self._local_collector.map.insert(package_hash, path)
+        await self._local_collector.indexer.register_package(package_hash, path, self._local_collector)
+
+    def _push_package(self, path_str: str) -> None:
+        run(self._push_package_async(path_str))
```

Comment:
Die Event-Handler werden offenbar in einem anderen Thread ausgeführt. Du musst daher mittels `asyncio.run_coroutine_threadsafe` die Coroutine in unserer main loop ausführen lassen.

Die `_pop_package` Methode darf ebenfalls nicht einfach so aus einem anderen Thread im Indexer Änderungen machen. Verwende dort am besten auch `asyncio.run_coroutine_threadsafe`.